### PR TITLE
allow submenu history preservation

### PR DESCRIPTION
**Pull request by albertored**
I added the possibility of having history preservation in a submenu.

When `AddSubMenu` is initialized with the `persistent_history_file` arg the logic is the following:
1. on entering the submenu the current history (from main app) is cleared and saved in `history` variable
2. history is read from the given  `persistent_history_file`
3. on exit the submenu history is written on  `persistent_history_file`
4. the main app saved history is pushed back in main app

Points 1 and 4 are exactly the same of before, I just changed their position in the code.
Points 2 and 3 are not performed if  `persistent_history_file` is not given.

**Open points / caveats**
* `persistent_history_file` for submenus  is an argument of `AddSubMenu` and not of `Cmd`, it may lead to confusion
* if  `persistent_history_file` is the same of main app then history on submenu is the one of main app, maybe we can add a check but still we will not be able to check that is different from all others submenus. Maybe is better to leave it as user responsibility
* I don't know how to test it programatically but I tried it in person and it seems to work well

### CODE REVIEW

**Comment by tleonhardt** [comment]
@albertored 
Thanks for the PR!  I like the possibility of preserving history separately for the different submenus.

Your open points and caveats don't concern me.  I'm fine with making it user responsibility to make sure things don't conflict in their application.

Could you merge in the latest changes from master.  This PR conflicts slightly with your previous PR #306 .  I could resolve the conflicts, but I would prefer you to do it in a way you are happy with.

**Comment by albertored** [comment]
rebased :+1:

File: cmd2.py [hunk 1]
```diff
@@ -832,8 +832,9 @@ def __init__(self,
                  reformat_prompt="{super_prompt}>> {sub_prompt}",
                  shared_attributes=None,
                  require_predefined_shares=True,
+                 create_subclass=False,
```
Comment (tleonhardt) [inline cmd2.py]: Looks good.  Thanks for resolving merge conflicts!

**Comment by tleonhardt** [comment]
@albertored I invited you to join the cmd2 contributor's team.  That will allow you to create branches directly within the main repository so you don't need to work out of a fork.  This makes it easier to use the integrated CI tools and to collaborate with others on the project.

**Pull request merged**